190612 Eye tracking docs update - part 1. (#4865)
microsoft/MixedRealityToolkit-Unity: Documentation/EyeTracking/EyeTracking_ExamplesOverview.md

# Eye tracking examples in MRTK

> This page covers how to get quickly started with using eye tracking in MRTK by building on our provided [MRTK eye tracking example package](https://github.com/Microsoft/MixedRealityToolkit-Unity/tree/mrtk_release/Assets/MixedRealityToolkit.Examples/Demos/EyeTracking).

Reviewer: ```suggestion
This page covers how to get quickly started with using eye tracking in MRTK by building on our provided [MRTK eye tracking examples](https://github.com/Microsoft/MixedRealityToolkit-Unity/tree/mrtk_release/Assets/MixedRealityToolkit.Examples/Demos/EyeTracking).
```

Reviewer: Mainly because there's no separate package for eye tracking examples, they are part of the examples package.

Reviewer: I don't want to lead people down a path of asking "where's the eye tracking package?"

The demo includes a number of different use cases ranging from implicit eye-based activations to how to seamlessly combine information about what you are looking at with **voice** and **hand** input. 
This enables users to quickly and effortlessly select and move holographic content across their view simply by looking at a target and saying _'Select'_ or performing a hand gesture. 
The demos also include an example for eye-gaze-directed scroll, pan and zoom of text and images on a slate. 
Finally, an example is provided for recording and visualizing the user's visual attention on a 2D slate.
In the following, we will go into more detail what each of the different samples in the [MRTK eye tracking example package](https://github.com/Microsoft/MixedRealityToolkit-Unity/tree/mrtk_release/Assets/MixedRealityToolkit.Examples/Demos/EyeTracking) includes:

![List of eye tracking scenes](../Images/EyeTracking/mrtk_et_list_et_scenes.jpg)

We will start with a quick overview of what the individual eye tracking demo scenes are about.
The individual sample scenes will be [loaded additively](https://docs.unity3d.com/ScriptReference/SceneManagement.LoadSceneMode.Additive.html).

> In case, you wonder what that entails, we will explain further below how to set up and test the MRTK eye tracking demos.

Reviewer: Nit: This sentence doesn't really add that much.

Initially I thought that this sentence would tell me more about additive scene loading ("you wonder what this entails") but then the second half of the sentence tells a totally different story (we'll explain how to test the demos!)

It reminds me of something someone told me the other day when they said "today is a cloudy day, as a result I want chocolate"

Author: It's not completely unrelated. There have been many questions of the additive loading and I wanted to raise it early on that 1) Yes, we do additive scene loading and 2) If you don't know what that means (because a lot of people didn't) then don't worry, we'll get to it. 
But anyways... I rephrased it for now, but it's already outdated anyways with the new scene loading system.

The samples let you experience one of our new magical input capabilities: **Eye tracking**! 
The demo includes a number of different use cases ranging from implicit eye-based activations to how to seamlessly combine information about what you are looking at with **voice** and **hand** input. 
This enables users to quickly and effortlessly select and move holographic content across their view simply by looking at a target and saying _'Select'_ or performing a hand gesture. 
The demos also include an example for eye-gaze-directed scroll, pan and zoom of text and images on a slate. 
Finally, an example is provided for recording and visualizing the user's visual attention on a 2D slate.

In the following, we'll go into more detail what each of the different samples in the [eye tracking example package](https://github.com/Microsoft/MixedRealityToolkit-Unity/tree/mrtk_release/Assets/MixedRealityToolkit.Examples/Demos/EyeTracking) includes:

> ![List of eye tracking scenes](../Images/EyeTracking/mrtk_et_list_et_scenes.jpg)

Reviewer: When looking at the preview, some of these images aren't being loaded (they look to be broken). I think that this is just an artifact of how github previews work, but it may be worth looking at these changes in a local docfx build.

Just to verify that the images and layout are all showing as expected:
https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/DevDocGuide.html#building-docfx-locally

Basically you can see what it will look like for real once it's on the github.io pages. If you need any help with those steps lemme know.

Reviewer: Actually, looks like this seems fine - using the diff page apparently isn't a great indication of accuracy, so resolving this one.

Author: Thanks, Will!

### 1. Load EyeTrackingDemo-00-RootScene.unity
The *EyeTrackingDemo-00-RootScene* is the base (_root_) scene that has all the core MRTK components included.
This is the scene that you need to load first and from which you will run the eye tracking demos. 
It comes with a graphical scene menu that allows you to easily switch between the different eye tracking samples which will be [loaded additively](https://docs.unity3d.com/ScriptReference/SceneManagement.LoadSceneMode.Additive.html).

![Scene menu in eye tracking sample](../Images/EyeTracking/mrtk_et_scenemenu.jpg)

The root scene includes a few core components that will persist across the additively loaded scenes, such as the MRTK configured profiles and scene camera. 
The _MixedRealityBasicSceneSetup_ (see screenshot below) includes a script that will automatically load the referenced scene on startup. 
By default this is _EyeTrackingDemo-02-TargetSelection_.  

![Example for the OnLoadStartScene script](../Images/EyeTracking/mrtk_et_onloadstartscene.jpg)

### 2. Adding scenes to the Build menu
To load additive scenes during runtime, you must add these scenes to your _Build Settings -> Scenes in Build_ menu first.

> It is important that the root scene is shown as the first scene in the list:

Reviewer: Do you know what happens if it's not (i.e. can you mention the dialog that you just added, which will appear if this is violated?).

This can help guide people on the right path (i.e. if we have a section saying here's how you know if something is going wrong, and they may search for the going wrong text, and then be led to this section about how to do things right)

Author: It's a good point. I leave it like this for now, mainly because this will probably change with the new changes from Lars anyways. I would like to add the revised docs for his change in a different PR though.

### 3. Play the eye tracking samples in the Unity Editor
After adding the eye tracking scenes to the Build Settings and loading the _EyeTrackingDemo-00-RootScene_, there is one last thing you may want to check: Is the _'OnLoadStartScene'_ script that is attached to the _MixedRealityBasicSceneSetup_ GameObject enabled? This is to let the root scene know which demo scene to load first.

![Example for the OnLoad_StartScene script](../Images/EyeTracking/mrtk_et_onloadstartscene.jpg)

Let's roll! Hit _"Play"_!
You should see several gems appear and should see the scene menu at the top.

![Sample screenshot from the ET target select scene](../Images/EyeTracking/mrtk_et_targetselect.png)

You should notice a small semitransparent circle at the center of your game view. 
This acts as an indicator (cursor) of your _simulated eye gaze_: 
Simply press down the _right mouse button_ and move the mouse to change its position. 
When the cursor is hovering over the gems, you will notice that it will snap to the center of the currently looked at gem. 
This is a great way to test if events are triggered as expected when _"looking"_ at a target. 

> Please beware that the _simulated eye gaze_ via mouse control is a rather poor supplement to our rapid and unintentional eye movements. 

Reviewer: ```suggestion
Please beware that this is simulating eye gaze via mouse control is a rather poor supplement to our rapid and unintentional eye movements. 
```

It is great for testing the basic functionality though before iterating on the design by deploying it to the HoloLens 2 device.
Coming back to our eye tracking sample scene: The gem rotates as long as being looked at and can be destroyed by "looking" at it and ...
- Pressing _Enter_ (which simulates saying "select")

> - Actually saying _"select"_ into your microphone

Reviewer: ```suggestion
- Saying _"select"_ into your microphone
```
